graph/src/components/metrics/stopwatch.rs

```diff
@@ -0,0 +1,169 @@
+#[derive(Clone)]
+pub struct StopwatchMetrics {
+    disabled: Arc<AtomicBool>,
+    inner: Arc<Mutex<StopwatchInner>>,
+}
+
+impl StopwatchMetrics {
+    pub fn new(
+        logger: Logger,
+        subgraph_id: SubgraphDeploymentId,
+        registry: Arc<dyn MetricsRegistry>,
+    ) -> Self {
+        let mut inner = StopwatchInner {
+            total_counter: *registry
+                .new_counter(
+                    format!("{}_total_secs", subgraph_id),
+                    format!("total time spent syncing"),
+                    HashMap::new(),
+                )
+                .expect("failed to register total_secs prometheus counter"),
+            logger,
+            subgraph_id,
+            registry,
+            counters: HashMap::new(),
+            section_stack: Vec::new(),
+            timer: Instant::now(),
+        };
+
+        // Start a base section so that all time is accounted for.
+        inner.start_section("unknown".to_owned());
+
+        StopwatchMetrics {
+            disabled: Arc::new(AtomicBool::new(false)),
+            inner: Arc::new(Mutex::new(inner)),
+        }
+    }
+
+    pub fn start_section(&self, id: &str) -> Section {
+        let id = id.to_owned();
+        if !self.disabled.load(Ordering::SeqCst) {
+            self.inner.lock().unwrap().start_section(id.clone())
+        }
+        Section {
+            id,
+            stopwatch: self.clone(),
+        }
+    }
+
+    pub fn disable(&self) {
+        self.disabled.store(true, Ordering::SeqCst)
+    }
+
+    fn end_section(&self, id: String) {
+        if !self.disabled.load(Ordering::SeqCst) {
+            self.inner.lock().unwrap().end_section(id)
+        }
+    }
+}
+
+/// We want to account for all subgraph indexing time, based on "wall clock" time. To do this we
+/// break down indexing into _sequential_ sections, and register the total time spent in each. So
+/// that there is no double counting, time spent in child sections doesn't count for the parent.
+struct StopwatchInner {
+    logger: Logger,
+    subgraph_id: SubgraphDeploymentId,
+    registry: Arc<dyn MetricsRegistry>,
+
+    // Counter for the total time the subgraph spent syncing.
+    total_counter: Counter,
+
+    // Counts the seconds spent in each section of the indexing code.
+    counters: HashMap<String, Counter>,
+
+    // The top section is the one that's currently executing.
+    section_stack: Vec<String>,
+
+    // The timer is reset whenever a section starts or ends.
+    timer: Instant,
+}
+
+impl StopwatchInner {
+    fn record_and_reset(&mut self) {
+        if let Some(section) = self.section_stack.last() {
+            // Get or create the counter.
+            let counter = if let Some(counter) = self.counters.get(section) {
+                counter.clone()
+            } else {
+                let name = format!("{}_{}_secs", self.subgraph_id, section);
+                let help = format!("indexing section {}", section);
```

Comment:
Here also I wonder if `indexing` isn't tying this too closely to indexing even though we may want to use stopwatches in other places as well.

Reply:
e.g. query execution

Reply:
true, I've removed the `indexing`.